### ISSUE TITLE
rpc: Advertise per-locality hosts

### ISSUE THREAD

**Issue by bdarnell** [cockroachdb/cockroach#25636]
In some deployment patterns (including [this multi-cloud example](https://www.cockroachlabs.com/blog/multi-cloud-deployment/), nodes have multiple network interfaces. This typically involves a "local" or "private" interface that is only accessible by a subset of the nodes, and a "global" or "public" interface that is slower or more expensive but accessible by any node. We want to route traffic over the local interface whenever possible.

The "right" answer here is to use DNS, with resolvers in each datacenter returning the best IP address that is reachable from that location. However, getting this set up correctly is non-trivial and has additional moving pieces that must be maintained. A solution integrated into cockroach may be significantly easier to deploy.

Strawman proposal: Add a new flag `--advertise-alternate-host` which takes a list of locality/hostname pairs. This mapping would go into the gossiped node status, and each node would look through the list for a value that matched its locality and use that address to connect.

Example:

    cockroach start --certs-dir=... --locality=region=west --advertise-host=$PUBLIC_IP --alternate-advertise-host=region=west:$PRIVATE_IP

**Comment by tschottdorf** [comment]
Have you considered extending the `--advertise-host` flag instead?

`--advertise-host=$PUBLIC_IP,$PRIVATE_IP@region=west1,$ANOTHER_PRIVATE_IP@region=west2`

or

`--advertise-host=$PUBLIC_IP --advertise-host=$PRIVATE_IP@region=west1`

where `$PUBLIC_IP` is equivalent to `$PUBLIC_IP@` and boils down to an empty constraint list.

The impl refactor https://github.com/cockroachdb/cockroach/blob/d72a431d61f236cb21d1a0e7821d5efde8185bc1/pkg/kv/replica_slice.go#L49 and https://github.com/cockroachdb/cockroach/blob/fb1beac985b6a51123c7e3c20e73d735590698c0/pkg/storage/raft_transport.go#L118 and maybe others to use a shared helper that takes the node's locality in addition to its gossip and target node.
When you say "the gossiped node status", you mean the node descriptor, right? There's also a node status
The code changes seem to be mostly refactoring. There's a bunch of testing work to be done to make sure it actually works, which implies setting up an environment in which it can be (roach)tested. This includes metrics, and this is easier if we restrict to the special case of at most one private IP. Is this a reasonable restriction?

BTW, would the "right" way to do it via DNS work with our rpcContext caching?

**Comment by bdarnell** [comment]
> Have you considered extending the --advertise-host flag instead?

Yeah, that's an option too. It's just a matter of coming up with the syntax we want. A new flag has the advantage of being perfectly backwards-compatible. 

> When you say "the gossiped node status", you mean the node descriptor, right? There's also a node status

Yeah, I meant the descriptor.

> BTW, would the "right" way to do it via DNS work with our rpcContext caching?

The caching is all in-process, and nodes can't change their locality, so this would work as long as your mapping is static (which is true for the CLI flags too).
